### PR TITLE
Implement step and compute_state functions (#2, #3)

### LINKED ISSUE #2: Implement the step function
Implement the step()-function of the environment according to the specification of the base paper. Depends on #3

### LINKED ISSUE #3: Implement compute_state function
Compute the state from the image and the action history. The base paper uses a pre-trained convolutional network as a feature extractor, however that network is not available so we have to come up with something by ourselves. Perhaps one of the pre-trained models Chainer provides?

### PR DESCRIPTION
**Pull request by J-L-O**
Closes #2, closes #3

I tried to stick to the base paper as closely as possible. The main deviation is that this environment uses the VGG16 network for feature extraction, since the network used in the base paper was not available.

### CODE REVIEW

File: environment.py [hunk 1]
```diff
@@ -1,42 +1,165 @@
...
+        self.action_set = {0: self.up,
+                           1: self.down,
+                           2: self.left,
+                           3: self.right,
+                           4: self.zoom_in,
+                           5: self.zoom_out,
+                           6: self.wider,
+                           7: self.taller,
+                           8: self.trigger
+                           }
```
Comment (bakoe) [inline environment.py:31]: Any reason why you (a) ordered and (b) named the actions different than in the paper?

image

If there's no specific reason, I'd prefer the following (a) order and (b) naming:

```python
        self.action_set = { 0: self.right,
                            1: self.left,
                            2: self.up,
                            3: self.down,
                            4: self.bigger,
                            5: self.smaller,
                            6: self.fatter,
                            7: self.taller,
                            8: self.trigger }
```

This may, to my mind, help us to debug and/or understand the ML results later on, since we can then directly switch back and forth between the algorithm and the base paper without having to change the order and naming in our minds continuously. :wink:

File: environment.py [hunk 2]
```diff
@@ -1,42 +1,165 @@
...
+
+        # self.feature_extractor.to_gpu(0)
```
Comment (bakoe) [inline environment.py]: I assume that commenting in this line would allow us to enable GPU support, at least for the VGG16 feature extractor? 
- If yes, I'd prefer to **either** have a comment describing that "Enabling the following line will enable GPU support for the feature extractor used" **or** even have a corresponding flag argument in the constructor that allows to enable the GPU support from outside (the latter would be my personal preference). 
- If not, I'd prefer to have the line removed for now, since commented-out code is considered bad practice imho since it could irritate either us or others reading the code.

File: environment.py [hunk 3]
```diff
@@ -1,42 +1,165 @@
...
+from PIL.Image import LANCZOS
+import numpy as np
 
 
 class TextLocEnv(gym.Env):
 
+    HISTORY_LENGTH = 10
+
     def __init__(self, image, true_bboxes):
         self.alpha = 0.2
```
Comment (bakoe) [inline environment.py]: Since you made the HISTORY_LENGTH a constant, I'd prefer making the alpha value a constant as well, i. e. by having `ALPHA = 0.2` in the `TextLocEnv`.

File: environment.py [hunk 4]
```diff
@@ -1,42 +1,165 @@
...
+                           }
+
+        # self.feature_extractor.to_gpu(0)
 
         self.image = image
         self.true_bboxes = true_bboxes
         self.history = []
-        self.bbox = (0, 0, image.width, image.height)
+        self.bbox = np.array([0, 0, image.width, image.height])
+        self.iou = 0
```
Comment (bakoe) [inline environment.py:40]: I'm not sure whether the abbreviation `iou` is the best name for this variable since it may not immediately be understandable. What about using `intersection_over_union` instead? 

However, that's just my personal opinion – using `iou` could be perfectly fine as well.

File: environment.py [hunk 5]
```diff
@@ -1,42 +1,165 @@
...
             state - the next state,
             reward - the reward,
             done - whether a terminal state was reached,
             info - any additional info"""
+        assert self.action_space.contains(action), "%r (%s) is an invalid action" % (action, type(action))
+
+        self.action_set[action]()
+
+        new_iou = self.compute_best_iou()
+        reward = np.sign(new_iou - self.iou)
```
Comment (bakoe) [inline environment.py]: According to the base paper, the reward for the terminal action `trigger` has to be calculated in a special way:

image
image

It seems like you did not implement this special reward calculation yet? Maybe we should include it in this PR or create a corresponding follow-up issue?

File: environment.py [hunk 6]
```diff
@@ -1,42 +1,165 @@
...
         """Reset the environment to its initial state (the bounding box covers the entire image"""
         self.history = []
-        self.bbox = (0, 0, self.image.width, self.image.height)
+        self.bbox = np.array([0, 0, self.image.width, self.image.height])
         self.state = self.compute_state()
+        self.done = False
 
         return self.state
 
     def render(self, mode='human'):
```
Comment (bakoe) [inline environment.py:158]: Just a question on the side: What is the `mode` argument for? I know that it has been part of the environment skeleton already … however, I'm still curious. 🙂

File: environment.py [hunk 7]
```diff
@@ -1,42 +1,165 @@
...
         """Render the current state"""
         copy = self.image.copy()
         draw = ImageDraw.Draw(copy)
-        draw.rectangle(self.bbox, outline=(255, 255, 255))
+        draw.rectangle(self.bbox.tolist(), outline=(255, 255, 255))
         copy.show()
 
+    def get_warped_bbox_contents(self):
+        croppped = self.image.crop(self.bbox)
```
Comment (bakoe) [inline environment.py]: Are you sure that this line works, i. e. that the `image.crop()` call succeeds when a np.array like `self.bbox` is passed?

image
— https://pillow.readthedocs.io/en/3.1.x/reference/Image.html#PIL.Image.Image.crop

According to [this explanation](https://medium.com/xster-tech/how-to-convert-numpy-array-to-tuple-aedbb6555dcc), we might need to use the following syntax instead:

```suggestion
        croppped = self.image.crop(tuple(self.bbox))
```

File: environment.py [hunk 8]
```diff
@@ -1,42 +1,165 @@
...
+        return croppped.resize((224, 224), LANCZOS)
+
     def compute_state(self):
+        return self.extract_features(), self.history
+
+    def extract_features(self):
         """Compute the state from the image, the bounding box and the action history"""
```
Comment (bakoe) [inline environment.py]: At the moment, the action history isn't considered in the state calculation, is it?

Comment (bakoe) [inline environment.py]: This comment was accidentally moved from `def compute_state(self):` to this function (`def extract_features(self):`), wasn't it?

File: environment.py [hunk 9]
```diff
@@ -1,42 +1,165 @@
...
+        new_iou = self.compute_best_iou()
+        reward = np.sign(new_iou - self.iou)
+        self.iou = new_iou
+
+        self.history.insert(0, self.to_one_hot(action))
+
+        if len(self.history) > TextLocEnv.HISTORY_LENGTH:
+            self.history.pop()
+
+        self.state = self.compute_state(), self.history
```
Comment (bakoe) [inline environment.py]: I'm a little bit confused since both this line and the `compute_state()` function from lines 151..152 …

```python
    def compute_state(self):
         return self.extract_features(), self.history
```

do return the `self.history` value. Isn't `self.state` then containing something like the following?

```
((<return value from extract_features()>, self.history), self.history)
```

File: environment.py [hunk 10]
```diff
@@ -1,42 +1,165 @@
...
+
+    def trigger(self):
+        self.done = True
+
+    def adjust_bbox(self, directions):
+        ah = round(self.alpha * (self.bbox[3] - self.bbox[1]))
+        aw = round(self.alpha * (self.bbox[2] - self.bbox[0]))
+
+        adjustments = np.array([aw, ah, aw, ah])
+        delta = directions * adjustments
```
Comment (bakoe) [inline environment.py:139]: Pretty cool usage of "Boolean" multiplication 🙂

File: environment.py [hunk 11]
```diff
@@ -1,42 +1,165 @@
...
+
+        new_box = self.bbox + delta
+        new_box[0] = max(new_box[0], 0)
+        new_box[1] = max(new_box[1], 0)
+        new_box[2] = min(new_box[2], self.image.width)
+        new_box[3] = min(new_box[3], self.image.height)
```
Comment (bakoe) [inline environment.py:145]: Shouldn't both this line and the line above constrain the bottom right co-ordinate of the new bounding box to `self.image.width - 1` and `self.image.height - 1`? Otherwise, we may end up with invalid bounding box co-ordinates when an image is e. g. 256 x 256 pixels large and the bottom right co-ordinate of the new bounding box is put to something like `(256, 20)`, which would be outside of the images' boundaries.

File: environment.py [hunk 12]
```diff
@@ -1,42 +1,165 @@
...
+        self.adjust_bbox(np.array([-1, 0, -1, 0]))
+
+    def right(self):
+        self.adjust_bbox(np.array([1, 0, 1, 0]))
+
+    def zoom_in(self):
+        self.adjust_bbox(np.array([1, 1, -1, -1]))
+
+    def zoom_out(self):
+        self.adjust_bbox(np.array([-1, -1, 1, 1]))
```
Comment (bakoe) [inline environment.py:123]: I'm not super sure whether the bigger and smaller actions (or zoom_in and zoom_out) should have "twice" the total effect (or "impact") as the other actions (which they currently do have – to my mind – due to the current implementation using an absolute sum of actions of 4 instead of 2, as used by the other actions)[1]. I did not find anything related in the base paper, unfortunately. 😢 Did you find something related that suggests using one of the approaches? 

[1] `zoom_out`: abs(-1) + abs(-1)+ abs(1) + abs(1) = 4,
`right`: abs(1) + abs(0) + abs(1) + abs(0) = 2

File: environment.py (hunk 7 above)
Comment (bakoe) [inline environment.py]: Typo:

```suggestion
        cropped = self.image.crop(self.bbox)
```

File: environment.py [hunk 13]
```diff
@@ -1,42 +1,165 @@
...
         """Render the current state"""
         copy = self.image.copy()
         draw = ImageDraw.Draw(copy)
-        draw.rectangle(self.bbox, outline=(255, 255, 255))
+        draw.rectangle(self.bbox.tolist(), outline=(255, 255, 255))
         copy.show()
 
+    def get_warped_bbox_contents(self):
+        croppped = self.image.crop(self.bbox)
+        return croppped.resize((224, 224), LANCZOS)
```
Comment (bakoe) [inline environment.py]: (Typo; see above)

```suggestion
        return cropped.resize((224, 224), LANCZOS)
```

File: environment.py [hunk 14]
```diff
@@ -1,42 +1,165 @@
...
     def compute_state(self):
+        return self.extract_features(), self.history
+
+    def extract_features(self):
         """Compute the state from the image, the bounding box and the action history"""
-        return 0
+        warped = self.get_warped_bbox_contents()
+        feature = self.feature_extractor.extract([warped], layers=["fc7"])["fc7"]
+
+        return feature[0]
```
Comment (bakoe) [inline environment.py:177]: I'm not sure about the syntax here – are you sure that `feature_extractor.extract(…)["fc7"][0]` returns the feature representation we're looking for?

**Review by bakoe — CHANGES_REQUESTED** [review]
Very nice, thanks for doing this!

I only had some minor complaints, as stated in the review comments.

Apart from that, on a high level, from my point of view the last missing pieces are:

- the special reward calculation necessary for the "trigger" action (see corresponding review comment)
- the "crossing out" (IoR) mechanism invoked by the "trigger" action (see #6 for a follow-up issue I already created for this)

What do you think, shall we include the special reward calculation for the "trigger" action in this PR or should we create a follow-up issue for this as well (similar to #6)?

File: environment.py (hunk 1 above)
Comment (J-L-O) [reply environment.py:31]: Makes sense, I'm going to change it

File: environment.py (hunk 2 above)
Comment (J-L-O) [reply environment.py]: The flag solution is definitely the best, yes

File: environment.py (hunk 3 above)
Comment (J-L-O) [reply environment.py]: I agree

File: environment.py (hunk 4 above)
Comment (J-L-O) [reply environment.py:40]: I was kind of unsure about that, `intersection_over_union` is a bit long imo, but it might still be better, yes

File: environment.py (hunk 5 above)
Comment (J-L-O) [reply environment.py]: I also realized that later, it's already fixed

File: environment.py (hunk 3 above)
Comment (bakoe) [reply environment.py]: (does apply to the newly introduced constants (namely tau – τ and eta – η) as well 🙂)

Comment (bakoe) [reply environment.py]: Actually, I'd really like having a one-line comment that (a) explains the semantic and (b) uses the original greek letter as in the base paper for the values of ALPHA, TAU and ETA:

```python
# ⍺: factor relative to the current box size that is used for every transformation action
ALPHA = 0.2
```
etc.

File: environment.py (hunk 6 above)
Comment (J-L-O) [reply environment.py:158]: I don't really get it either, most predefined environments ignore the argument just like we do. The atari_env offers two modes, so you might want to look into that if you want find out more

File: environment.py (hunk 4 above)
Comment (bakoe) [reply environment.py:40]: Hmm, it's really quite long, especially concerning that we have multiple methods as well which could/should use the long name as part of their function name then as well.

Maybe we should just stick to `iou` then – I don't know.

If we had doc comments to the functions, we could easily explain the meaning of `iou` there as well, e. g. in a doc comment for the `compute_best_iou()` method. Hmm … 😉

File: environment.py (hunk 7 above)
Comment (J-L-O) [reply environment.py]: Well, it definitely works right now, however your suggestion might be cleaner

File: environment.py (hunk 8 above)
Comment (J-L-O) [reply environment.py]: Yep

File: environment.py (hunk 7 above)
Comment (bakoe) [reply environment.py]: As we are at "it definitely works right now" – is there any way to actually test out the environment right now? Like some sort of debug/interactive mode for OpenAI gym environments that one could use to manually trigger the actions and e. g. continuously render the resulting image?

File: environment.py (hunk 12 above)
Comment (J-L-O) [reply environment.py:123]: No, the paper doesn't say anything about that. I guess we should just use whatever turns out to give better results

File: environment.py (hunk 13 above)
Comment (J-L-O) [reply environment.py]: Fixed

File: environment.py (hunk 11 above)
Comment (J-L-O) [reply environment.py:145]: The `Image.crop()` function excludes the last pixel, if you take a 256x256 image and say `img.crop((0, 0, 255, 255))` you cut of the last row/column

File: environment.py (hunk 7 above)
Comment (J-L-O) [reply environment.py]: Right now I just manually call the step() and render() functions.

File: environment.py (hunk 9 above)
Comment (J-L-O) [reply environment.py]: True, fixed

**Review by bakoe — APPROVED** [review]

**Comment by bakoe** [comment]
Thanks for integrating the review feedback, @J-L-O! From my point of view, the PR is ready to merge! 

PS: We may use "Squash and Merge" in order to keep the commit history of the master branch somewhat clean:

image
